### DISCUSSION TITLE
Confusing history behavior with diffput / diffget

### DISCUSSION THREAD

**Discussion by shmerl** [neovim/neovim#29602] [Q&A] [answered]
I noticed something strange about neovim's diff mode.

* Let's say I open a diff between two files

```
nvim -d a b
```
This opens a split between `a` and `b`, with `a` being the active window.

* Then let's say I do a couple of `diffput` from `a` which updates `b`.

When I switch to `b` and do Undo (`u`) it undoes both changes, not just the last one. Is that expected?

Strangely, if between first and second diffput I switch to `b` temporarily and then back to `a`, then in the end dong undo in b does it in steps like I'd expect (i.e. first the second change).

This feels like a bug, unless I'm missing something? I can open a bug if it is such.

**Comment by zeertzjq** [comment]
Cannot reproduce

**Comment by shmerl** [comment]
Interesting. I have just a few plugins, but may be something is interfering? I can try testing without any plugins to confirm.

**Reply by zeertzjq** [reply]
It will also be better if you share the contents of the files `a` and `b`.

**Reply by shmerl** [reply]
Hmm, I can't reproduce it when doing this

```
nvim -u NONE -d a b
```
a contains:

```
aaa
bbb
ccc
```

b contains
```
aa1
bbb
cc2
```

I'll try to narrow down what can be interfering with it.

**Comment by zeertzjq** [comment]
- Does this happen if you press `dp` instead of using `:diffput`?
- Do you have `<CR>` mapped in cmdline mode?

**Reply by shmerl** [reply]
hm, `dp` in normal mode isn't affecting it at all for some reason in my custom configuration. I don't recall ever overriding it.

**Reply by shmerl** [reply]
I do have such kind of mapping:

```
vim.keymap.set('c', '<C-R>',
  function()
    return ((vim.fn.getcmdline() == '') and (vim.fn.getcmdtype() == ':')) and 'FzfLua command_history<CR>' or '<C-R>'
  end,
  { expr = true }
)
```

**Reply by zeertzjq** [reply]
That's `<C-R>`, which isn't involved here.

**Reply by shmerl** [reply]
Let me turn off all my mappings and plugins one by one until I narrow it down.

**Reply by shmerl** [reply]
OK, `dp` wasn't working for me becasue I remapped `p` to `gP`. Commenting that out, when using `dp` history works properly indeed. But when using `:diffput` it still doesn't work i.e. undo undoes all changes.

Checking what introduces regression for `:diffput`.

**Reply by shmerl** [reply]
OK, I found the offending plugin, it's this one: https://github.com/henrik/vim-indexed-search with it disabled, undo works correctly.

**Reply by shmerl** [reply]
I guess I can look into fixing it or avoid using it and find / make better replacement. I might not even need it since stock search shows index correctly (may be in the past it didn't?).

**Comment by shmerl** [comment]
That helped explicit `:diffget` / `:diffput`, thanks. But now I have a problem with custom diff command which I can reproduce even with no custom configs.

Let's say the same scenario as above, but add such kind of mapping:

```
:lua vim.keymap.set('n', '>', function() return vim.o.diff and '<Cmd>diffput<CR>' or '>' end, { expr = true })
```

Then let's say try to do diffput by pressing `>` for a couple of different lines. It works, but undo is now for the whole change.

Can this mapping be improved somehow to handle history correctly?

**Reply by zeertzjq** [reply]
`:h undo-break`

**Reply by shmerl** [reply]
Hm, I'm not clear how to use it. I need to enter insert mode and feed `<i_CTRL-G_u>` to it? I.e. how do I modify my expr mapping to do that?

**Reply by shmerl** [reply]
I tried replacing `"<Cmd>diffput<CR>"` with `"<Cmd>diffput<CR>i<i_CTRL-G_u>"` or `"<Cmd>diffput<CR><i_CTRL-G_u>"` but that didn't affect it.

Also, I suppose that break has to be applied to the non focused window somehow for diffput, right? I guess I'd need to modify it to temporarily change focus, insert that and then change it back.

Even focus jump itself makes history work though, as above.

**Reply by zeertzjq** [reply]
Right, it needs to use the target buffer. Then maybe a better may is to use `t` flag of `feedkeys()`, i.e. changing the mapping RHS to `<Cmd>call feedkeys('dp', 'tnix')<CR>`.

**Reply by shmerl** [reply]
Hm, since dp doesn't work for me, falling back to `:diffput` with that didn't really work. Plus I have a more complicated visual selection command that's more Lua based like this:

```
vim.keymap.set('v', '<', function() return vim.o.diff and "<Cmd>lua visual_diff_get()<CR>" or '<' end, { expr = true })
vim.keymap.set('v', '>', function() return vim.o.diff and "<Cmd>lua visual_diff_put()<CR>" or '>' end, { expr = true })
```

Where

visual_diff_put is something like this:

```
function visual_diff_put()
  vim.cmd.diffput({ range = require("helpers/neovim").get_visual_selection_line_range() })
end
```

So I still need to figure out how to break history block explicitly I suppose. I think there should be some programmatic call that tells neovim "break history" instead of this insert hack.

**Reply by zeertzjq** [reply, marked as answer]
Yes, that's several lines below:
```
Setting the value of 'undolevels' also closes the undo block.  Even when the
new value is equal to the old value.  Use `g:undolevels` to explicitly read
and write only the global value of 'undolevels'. >
	let &g:undolevels = &g:undolevels
```
However, it isn't straightforward to find the diff buffer. Or maybe you can just use `nvim_win_call` to do `vim.bo.undolevels = vim.bo.undolevels` for every window in the current tabpage?

**Reply by shmerl** [reply]
Yeah, that's what I was just wondering, how do I find the diff buffer. I can try doing it for all windows, yeah. Though it feels like a crude workaround.

**Reply by shmerl** [reply]
OK, this worked in the end even though it feels hackish. Thanks a lot for the tips:

```lua
-- breaks history block for all windows in the current tab
-- needed for diff put operations due to target buffer not breaking history block otherwise
function break_history_block()
  local current_window_id = vim.api.nvim_tabpage_get_win(0)
  for _, window_id in pairs(vim.api.nvim_tabpage_list_wins(0)) do
    if (window_id ~= current_window_id) then
      vim.api.nvim_win_call(window_id, function() if (vim.o.diff) then vim.bo.undolevels = vim.bo.undolevels end end)
    end
  end

  return true
end

-- diff merging with < > for whole lines in normal mode
vim.keymap.set('n', '<', function() return vim.o.diff and '<Cmd>diffget<CR>' or '<' end, { expr = true })
vim.keymap.set('n', '>', function() return vim.o.diff and break_history_block() and '<Cmd>diffput<CR>' or '>' end, { expr = true })
```

**Reply by shmerl** [reply]
By the way, is it better to use `vim.o.undolevels = vim.o.undolevels` , not `vim.bo.undolevels = vim.bo.undolevels` in the above?

Help recommends using the global one pointing out that setting local one to itself can mess up the global one (though when I tested it I didn't observe such mess happening, i.e. when using `vim.bo.undolevels = vim.bo.undolevels`, `vim.o.undolevels` wasn't modified. Both work in practice to solve the problem.

**Reply by zeertzjq** [reply]
No, `vim.o` should not be used, and is exactly what the help recommends against. Either `vim.bo` or `vim.go` should be fine.
